### ISSUE TITLE
repilcating experiments

### ISSUE THREAD

**Issue by immersky** [key2miao/TSTNN#12]
Dear Mr. Wang,

I tried to replicate TSTNN: TWO-STAGE TRANSFORMER BASED NEURAL NETWORK FOR SPEECH  ENHANCEMENT IN THE TIME DOMAIN.[Kai Wang, Bengbeng He, Wei-Ping Zhu], but there exists some problems.

I used `gen_pair.py` to generate the  pairs from the dataset with 28 speakers, and I didn't change any other codes in this repository. However after 150 epoches on RTX3090 with batch_size 2, the results of the best-model in `test.py` showed worse performance compared to that in the dissertation. The results are as follows.
Result1// loss_time + 0.6 * loss_freq
|       | REPLICATING | ORIGINAL(IN DISSERTATION) |
| ----- | ----------- | ------------------------- |
| STOI: | 0.9450      | 0.95                      |
| SSNR: | 9.2971      | 9.70                      |
| PESQ: | 2.8254      | 2.96                      |
| CSIG: | 4.2356      | 4.33                      |
| CBAK: | 3.4413      | 3.53                      |
| COVL: | 3.5462      | 3.67                      |

Result2 //loss = 0.8 * loss_time + 0.2 * loss_freq
|       | REPLICATING | ORIGINAL(IN DISSERTATION) |
| ----- | ----------- | ------------------------- |
| STOI: | 0.9478  | 0.95                      |
| SSNR: | 9.4876   | 9.70                      |
| PESQ: | 2.9128   | 2.96                      |
| CSIG: |  4.2624    | 4.33                      |
| CBAK: | 3.4989  | 3.53                      |
| COVL:  | 3.6072  | 3.67                      |

Do you have any advice/reference on how to approach the best performance?
Also I'm confused about the parameter of loss_time. Should it be 'loss = 0.4 * loss_time + 0.6 * loss_freq'  or  'loss = 0.8 * loss_time + 0.2 * loss_freq'?

> ![ttttttt](https://user-images.githubusercontent.com/74494790/190938827-e3dd34de-3425-410d-9932-d8c20b4fb6b2.png)

Best regards

**Comment by key2miao** [comment]
> Dear Mr. Wang,
> 
> I tried to replicate TSTNN: TWO-STAGE TRANSFORMER BASED NEURAL NETWORK FOR SPEECH ENHANCEMENT IN THE TIME DOMAIN.[Kai Wang, Bengbeng He, Wei-Ping Zhu], but there exists some problems.
> 
> I used `gen_pair.py` to generate the pairs from the dataset with 28 speakers, and I didn't change any other codes in this repository. However after 150 epoches on RTX3090 with batch_size 2, the results of the best-model in `test.py` showed worse performance compared to that in the dissertation. The results are as follows. Result1// loss_time + 0.6 * loss_freq
> 
> REPLICATING	ORIGINAL(IN DISSERTATION)
> STOI:	0.9450	0.95
> SSNR:	9.2971	9.70
> PESQ:	2.8254	2.96
> CSIG:	4.2356	4.33
> CBAK:	3.4413	3.53
> COVL:	3.5462	3.67
> Result2 //loss = 0.8 * loss_time + 0.2 * loss_freq
> 
> REPLICATING	ORIGINAL(IN DISSERTATION)
> STOI:	0.9478	0.95
> SSNR:	9.4876	9.70
> PESQ:	2.9128	2.96
> CSIG:	4.2624	4.33
> CBAK:	3.4989	3.53
> COVL:	3.6072	3.67
> Do you have any advice/reference on how to approach the best performance? Also I'm confused about the parameter of loss_time. Should it be 'loss = 0.4 * loss_time + 0.6 * loss_freq' or 'loss = 0.8 * loss_time + 0.2 * loss_freq'?
> 
> > ![ttttttt](https://user-images.githubusercontent.com/74494790/190938827-e3dd34de-3425-410d-9932-d8c20b4fb6b2.png)
> 
> Best regards

Hello, thank you very much for reaching out. The 'best model' means the checkpoint with the lowest loss during the training rather than the one with best evaluation results. As you could know, in speech enhancement, the loss performane is not consistent with denoising performance. You should try other checkpoints (maybe have larger loss value but having better evaluation performance). I hope my answer can solve your problem.

**Comment by immersky** [comment]
> > Dear Mr. Wang,
> > I tried to replicate TSTNN: TWO-STAGE TRANSFORMER BASED NEURAL NETWORK FOR SPEECH ENHANCEMENT IN THE TIME DOMAIN.[Kai Wang, Bengbeng He, Wei-Ping Zhu], but there exists some problems.
> > I used `gen_pair.py` to generate the pairs from the dataset with 28 speakers, and I didn't change any other codes in this repository. However after 150 epoches on RTX3090 with batch_size 2, the results of the best-model in `test.py` showed worse performance compared to that in the dissertation. The results are as follows. Result1// loss_time + 0.6 * loss_freq
> > REPLICATING	ORIGINAL(IN DISSERTATION)
> > STOI:	0.9450	0.95
> > SSNR:	9.2971	9.70
> > PESQ:	2.8254	2.96
> > CSIG:	4.2356	4.33
> > CBAK:	3.4413	3.53
> > COVL:	3.5462	3.67
> > Result2 //loss = 0.8 * loss_time + 0.2 * loss_freq
> > REPLICATING	ORIGINAL(IN DISSERTATION)
> > STOI:	0.9478	0.95
> > SSNR:	9.4876	9.70
> > PESQ:	2.9128	2.96
> > CSIG:	4.2624	4.33
> > CBAK:	3.4989	3.53
> > COVL:	3.6072	3.67
> > Do you have any advice/reference on how to approach the best performance? Also I'm confused about the parameter of loss_time. Should it be 'loss = 0.4 * loss_time + 0.6 * loss_freq' or 'loss = 0.8 * loss_time + 0.2 * loss_freq'?
> > > ![ttttttt](https://user-images.githubusercontent.com/74494790/190938827-e3dd34de-3425-410d-9932-d8c20b4fb6b2.png)
> > 
> > 
> > Best regards
> 
> Hello, thank you very much for reaching out. The 'best model' means the checkpoint with the lowest loss during the training rather than the one with best evaluation results. As you could know, in speech enhancement, the loss performane is not consistent with denoising performance. You should try other checkpoints (maybe have larger loss value but having better evaluation performance). I hope my answer can solve your problem.

Dear Mr. Wang,

According to your advice, I tested the model which showed higher SNR when training, and it showed better performance which could be comparable to the original one. 

|  |REPLICATING(epoch 178) | ORIGINAL(IN DISSERTATION)       |
| ---------------------- | ------------------------- | ---- |
| STOI:                  | 0.9469                    | 0.95 |
| SSNR:                  | 9.6960                    | 9.70 |
| PESQ:                  | 2.9564                    | 2.96 |
| CSIG:                  | 4.3351                    | 4.33 |
| CBAK:                  | 3.5335                    | 3.53 |
| COVL:                  | 3.6682                    | 3.67 |

Thank you so much for taking the time!

**Comment by Litchi1999** [comment]
hi！Can I ask you what is the input about test.py?
